### PR TITLE
Refactored `regrid()` and removed unnecessary code not needed anymore due to new iris version

### LINKED ISSUE #1802: Remove special code for unstructured nearest regridder as bug in iris has been fixed
In iris v3.4, a bug will be fixed that lead to wrong masks and `dtype` of data regridded with our `unstructured_nearest` regridder (https://github.com/SciTools/iris/issues/4463). For this reason, special code that has been added to account for this bug can be removed:

https://github.com/ESMValGroup/ESMValCore/blob/89fbe934ff96b446eca8a4f33bafa2a7f614cc0d/esmvalcore/preprocessor/_regrid.py#L617-L646

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR slightly refactors `regrid()` and removes code regarding the `unstructured_nearest` scheme that is not necessary anymore since the release of iris 3.4.0.

 I also tested this with real data; everything seems to run fine :+1: 

Closes #1802 

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ dependencies:
   - geopy
   - humanfriendly
   - importlib_resources
-  - iris>=3.2.1
+  - iris>=3.4.0
```
Comment (valeriupredoi) [inline environment.yml:20]: are we sure we want to be **this** restrictive about the iris version? :beer:

Comment (schlunma) [reply environment.yml:20]: Yes, we do, this change was only introduced in iris 3.4 (see number 10 here: https://scitools-iris.readthedocs.io/en/latest/whatsnew/3.4.html#bugs-fixed). I will need iris 3.4 for the ICON to UGRID stuff (#1653) anyway...

Comment (valeriupredoi) [reply environment.yml:20]: mmm ok - given that I would be very happy to pin to the iris version that will have the fix for netCDF4>1.6.0 issue, I can understand this one too :+1: I'd still want to hear what the techy folks think, but by all means - fine by me - @ESMValGroup/technical-lead-development-team are we ready to start pinning and using only the latest versions of iris?

**Review by valeriupredoi — APPROVED** [review]

File: environment.yml (hunk 1 above)
Comment (zklaus) [reply environment.yml:20]: Looks like no objections.

**Pull request merged**